pkg/operator/ceph/object/controller.go

```diff
@@ -17,215 +17,329 @@ limitations under the License.
+		err = c.Watch(&source.Kind{Type: t}, &handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &cephv1.CephObjectStore{},
+		}, opcontroller.WatchPredicateForNonCRDObject(controllerTypeMeta))
 		if err != nil {
-			// This handles the case where the operator is running, the external cluster has been upgraded and a CR creation is called
-			// If that's a major version upgrade we fail, if it's a minor version, we continue, it's not ideal but not critical
-			logger.Errorf("refusing to run new crd. %v", err)
-			updateCephObjectStoreStatus(objectStore.GetName(), objectStore.GetNamespace(), k8sutil.FailedStatus, c.context)
-			return
+			return err
 		}
 	}
-	c.createOrUpdateStore(objectStore)
-	updateCephObjectStoreStatus(objectStore.GetName(), objectStore.GetNamespace(), k8sutil.ReadyStatus, c.context)
+
+	return nil
 }
 
-func (c *ObjectStoreController) onUpdate(oldObj, newObj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Updating object store for an external ceph cluster is disabled because no Ceph image is specified")
-		return
+// Reconcile reads that state of the cluster for a cephObjectStore object and makes changes based on the state read
+// and what is in the cephObjectStore.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileCephObjectStore) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// workaround because the rook logging mechanism is not compatible with the controller-runtime loggin interface
+	reconcileResponse, err := r.reconcile(request)
+	if err != nil {
+		logger.Errorf("failed to reconcile %v", err)
 	}
 
-	// if the object store spec is modified, update the object store
-	oldStore, err := getObjectStoreObject(oldObj)
+	return reconcileResponse, err
+}
+
+func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the cephObjectStore instance
+	cephObjectStore := &cephv1.CephObjectStore{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, cephObjectStore)
 	if err != nil {
-		logger.Errorf("failed to get old objectstore object. %v", err)
-		return
+		if kerrors.IsNotFound(err) {
+			logger.Debug("cephObjectStore resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrap(err, "failed to get cephObjectStore")
+	}
+
+	// The CR was just created, initializing status fields
+	if cephObjectStore.Status == nil {
+		cephObjectStore.Status = &cephv1.Status{}
+		cephObjectStore.Status.Phase = k8sutil.Created
+		err := opcontroller.UpdateStatus(r.client, cephObjectStore)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
```

Comment:
What if we just log the error when updating the status and allow the reconcile to continue? It doesn't seem like a blocking issue.

Reply:
It will if it fails and we continue then this `cephObjectStore.Status.Phase = k8sutil.ReadyStatus` will fail. Failing the reconcile is not a big deal IMHO, a few seconds after it'll reconcile again and most likely pass.